README.txt
RVideo

RVideo allows you to inspect and process video files.


Installation is a little involved. First, install the gem:

  sudo gem install rvideo
  
Next, install ffmpeg and (possibly) other related libraries. This is
documented elsewhere on the web, and can be a headache. If you are on OS X,
the Darwinports build is reasonably good (though not perfect). Install with:

  sudo port install ffmpeg

Or, for a better build (recommended), add additional video- and audio-related
libraries, like this:

  sudo port install ffmpeg +lame +libogg +vorbis +faac +faad +xvid +x264 +a52
  
Most package management systems include a build of ffmpeg, but many include a
poor build. So you may need to compile from scratch.

If you want to create Flash Video files, also install flvtool2:

  sudo gem install flvtool2

Once ffmpeg and RVideo are installed, you're set. 

To inspect a file, initialize an RVideo file inspector object. See the 
documentation for details.

A few examples:

  file = RVideo::Inspector.new(:file => "#{APP_ROOT}/files/input.mp4")
  
  file = RVideo::Inspector.new(:raw_response => @existing_response)
  
  file = RVideo::Inspector.new(:file => "#{APP_ROOT}/files/input.mp4",
                                :ffmpeg_binary => "#{APP_ROOT}/bin/ffmpeg")

  file.fps        # "29.97"
  file.duration   # "00:05:23.4"

To transcode a video, initialize a Transcoder object.

  transcoder = RVideo::Transcoder.new

Then pass a command and valid options to the execute method

  recipe = "ffmpeg -i $input_file$ -ar 22050 -ab 64 -f flv -r 29.97 -s"
  recipe += " $resolution$ -y $output_file$"
  recipe += "\nflvtool2 -U $output_file$"
  begin
    transcoder.execute(recipe, {:input_file => "/path/to/input.mp4",
      :output_file => "/path/to/output.flv", :resolution => "640x360"})
  rescue TranscoderError => e
    puts "Unable to transcode file: #{e.class} - #{e.message}"
  end

If the job succeeds, you can access the metadata of the input and output
files with:

  transcoder.original     # RVideo::Inspector object
  transcoder.processed    # RVideo::Inspector object

If the transcoding succeeds, the file may still have problems. RVideo
will populate an errors array if the duration of the processed video
differs from the duration of the original video, or if the processed
file is unreadable.

Thanks to Peter Boling for early work on RVideo.

Contribute to RVideo! If you want to help out, there are a few things you can 
do.

- Use, test, and submit bugs/patches
- We need a RVideo::Tools::Mencoder class to add mencoder support.
- Other tool classes would be great - On2, mp4box, Quicktime (?), etc.
- Submit other fixes, features, optimizations, and refactorings

If RVideo is useful to you, you may also be interested in RMovie, another Ruby
video library. See http://rmovie.rubyforge.org/ for more.

Finally, watch for Spinoza, a commercial video transcoder built by Slantwise 
Design. Spinoza uses RVideo for its video processing, but adds file queuing,  
distributed transcoding, a web-based transcoder dashboard, and more. See 
http://spinoza.tv or http://slantwisedesign.com for more.

Copyright (c) 2007 Jonathan Dahl and Slantwise Design. Released under the MIT 
license.

History.txt
== 0.9.4 2007-12-12

3 major enhancements:
* Changed transcoder interface. The preferred interface is now to pass the input file to RVideo::Transcoder.new. This allows you to create multiple output files from the same Transcoder object, and inspects the input file metadata before the job is transcoded.
* Added screengrab functionality to the Inspector class. Fire up an inspector instance on a file, and you can take one or more screenshots through inspected_file.capture_frame("50%").
* Added resolution/aspect support. Pass :width and :height parameters to a Transcoder instance, and include $resolution$ or $aspect_ratio$ in the recipe. 

4 minor enhancements:
* Remove old/unnecessary files and features
* Three additional ffmpeg results: unsupported codec and no output streams
* One additional flvtool2 result: empty input file
* Check that input file exists earlier in the transcoding process

== 0.9.3 2007-10-30

One minor enhancement:
* Reraise all unhandled RVideo exceptions as TranscoderErrors

== 0.9.2 2007-10-30

One minor bug fix:
* Correctly parse invalid files, where duration and bitrate are N/A, but start: 0.000 is included

== 0.9.1 2007-10-11

One major enhancement:
* Added Mencoder support. (Andre Medeiros)

Two minor enhancements: 
* Added total_time method to RVideo::Transcoder instances.
* Added another error condition for ffmpeg - codec not found.

Two notes:
* Tried and tested using open3 and open4 for command execution, but mencoder would unexpectedly hang with certain files when using these. Reverted these changes.
* Mencoder has basic unit tests, but needs more tests. In particular, example output should be added for a variety of cases (especially failures and errors).

== 0.9.0 2007-09-27

* Public RVideo release.

== 0.8.0 2007-09-27

* RVideo rewrite.


spec/units/flvtool2_spec.rb
require File.dirname(__FILE__) + '/../spec_helper'

module RVideo
  module Tools
  
    describe Flvtool2 do
      before do
        setup_flvtool2_spec
      end
      
      it "should initialize with valid arguments" do
        @flvtool2.class.should == Flvtool2
      end
      
      it "should have the correct tool_command" do
        @flvtool2.tool_command.should == 'flvtool2'
      end
      
      it "should call parse_result on execute, with a result string" do
        @flvtool2.should_receive(:parse_result).once.with /\AERROR: No such file or directory/
        @flvtool2.execute
      end
      
      it "should mixin AbstractTool" do
        Flvtool2.included_modules.include?(AbstractTool::InstanceMethods).should be_true
      end
      
      it "should set supported options successfully" do
        @flvtool2.options[:temp_file].should == @options[:temp_file]
        @flvtool2.options[:output_file].should == @options[:output_file]
      end
      
    end

    describe Flvtool2, " when parsing a result" do
      before do
        setup_flvtool2_spec
      end
      
      it "should set metadata if called with -P option" do
        @flvtool2.send(:parse_result, @metadata_result).should be_true
        @flvtool2.raw_metadata.should == @metadata_result
      end
      
      it "should succeed but not set metadata without -P option" do
        @flvtool2.send(:parse_result,"").should be_true
      end
    end
    
    context Flvtool2, " result parsing should raise an exception" do
      
      setup do
        setup_flvtool2_spec
      end
      
      specify "when not passed a command" do
        lambda {
          @flvtool2.send(:parse_result, @helptext)
        }.should raise_error(TranscoderError::InvalidCommand, /flvtool2 help text/)
      end
      
      specify "when receiving an empty file" do
        lambda {
          @flvtool2.send(:parse_result, @empty_file)
        }.should raise_error(TranscoderError::InvalidFile, /Output file was empty/)
      end
      
      specify "when passed an invalid input file" do
        lambda {
          @flvtool2.send(:parse_result, @non_flv_input)
        }.should raise_error(TranscoderError::InvalidFile, "input must be a valid FLV file")
      end
      
      specify "when input file not found" do
        lambda {
          @flvtool2.send(:parse_result, @no_input_file)
        }.should raise_error(TranscoderError::InputFileNotFound, /^ERROR: No such file or directory/)
      end
      
      specify "when receiving unexpected results" do
        lambda {
          @flvtool2.send(:parse_result, @unexpected_results)
        }.should raise_error(TranscoderError::UnexpectedResult, /ffmpeg/i)
      end      
    end
  end
end

def setup_flvtool2_spec
  @options = {:temp_file => "foo", :output_file => "bar"}
  @command = "flvtool2 -U $temp_file$ $output_file$"
  @flvtool2 = RVideo::Tools::Flvtool2.new(@command, @options)
  
  @helptext = "FLVTool2 1.0.6
  Copyright (c) 2005-2007 Norman Timmler (inlet media e.K., Hamburg, Germany)
  Get the latest version from http://www.inlet-media.de/flvtool2
  This program is published under the BSD license.

  Usage: flvtool2 [-ACDPUVaciklnoprstvx]... [-key:value]... in-path|stdin [out-path|stdout]

  If out-path is omitted, in-path will be overwritten.
  In-path can be a single file, or a directory. If in-path is a directory,
  out-path has to be likewise, or can be omitted. Directory recursion
  is controlled by the -r switch. You can use stdin and stdout keywords
  as in- and out-path for piping or redirecting.

  Chain commands like that: -UP (updates FLV file than prints out meta data)

  Commands:
    -A            Adds tags from -t tags-file
    -C            Cuts file using -i inpoint and -o outpoint
    -D            Debugs file (writes a lot to stdout)
    -H            Helpscreen will be shown
    -P            Prints out meta data to stdout
    -U            Updates FLV with an onMetaTag event

  Switches:
    -a            Collapse space between cutted regions
    -c            Compatibility mode calculates some onMetaTag values different
    -key:value    Key-value-pair for onMetaData tag (overwrites generated values)
    -i timestamp  Inpoint for cut command in miliseconds
    -k            Keyframe mode slides onCuePoint(navigation) tags added by the
                  add command to nearest keyframe position
    -l            Logs FLV stream reading to stream.log in current directory
    -n            Number of tag to debug
    -o timestamp  Outpoint for cut command in miliseconds
    -p            Preserve mode only updates FLVs that have not been processed
                  before
    -r            Recursion for directory processing
    -s            Simulation mode never writes FLV data to out-path
    -t path       Tagfile (MetaTags written in XML)
    -v            Verbose mode
    -x            XML mode instead of YAML mode

  REPORT BUGS at http://projects.inlet-media.de/flvtool2
  Powered by Riva VX, http://rivavx.com"
  
  @non_flv_input = "ERROR: IO is not a FLV stream. Wrong signature.
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flv/stream.rb:393:in `read_header'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flv/stream.rb:57:in `initialize'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:272:in `new'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:272:in `open_stream'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:238:in `process_files'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:225:in `each'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:225:in `process_files'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:44:in `execute!'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2.rb:168:in `execute!'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2.rb:228
  ERROR: /opt/local/lib/ruby/vendor_ruby/1.8/rubygems/custom_require.rb:27:in `gem_original_require'
  ERROR: /opt/local/lib/ruby/vendor_ruby/1.8/rubygems/custom_require.rb:27:in `require'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/bin/flvtool2:2
  ERROR: /opt/local/bin/flvtool2:18:in `load'
  ERROR: /opt/local/bin/flvtool2:18"
  
  @no_input_file = "ERROR: No such file or directory - /Users/jon/code/spinoza/rvideo/foobar
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:259:in `initialize'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:259:in `open'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:259:in `open_stream'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:238:in `process_files'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:225:in `each'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:225:in `process_files'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:44:in `execute!'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2.rb:168:in `execute!'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2.rb:228
  ERROR: /opt/local/lib/ruby/vendor_ruby/1.8/rubygems/custom_require.rb:27:in `gem_original_require'
  ERROR: /opt/local/lib/ruby/vendor_ruby/1.8/rubygems/custom_require.rb:27:in `require'
  ERROR: /opt/local/lib/ruby/gems/1.8/gems/flvtool2-1.0.6/bin/flvtool2:2
  ERROR: /opt/local/bin/flvtool2:18:in `load'
  ERROR: /opt/local/bin/flvtool2:18"
  
  @unexpected_results = "FFmpeg version CVS, Copyright (c) 2000-2004 Fabrice Bellard
  Mac OSX universal build for ffmpegX
    configuration:  --enable-memalign-hack --enable-mp3lame --enable-gpl --disable-vhook --disable-ffplay --disable-ffserver --enable-a52 --enable-xvid --enable-faac --enable-faad --enable-amr_nb --enable-amr_wb --enable-pthreads --enable-x264 
    libavutil version: 49.0.0
    libavcodec version: 51.9.0
    libavformat version: 50.4.0
    built on Apr 15 2006 04:58:19, gcc: 4.0.1 (Apple Computer, Inc. build 5250)

  Seems that stream 1 comes from film source: 600.00 (600/1) -> 59.75 (239/4)
  Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'jobjob2.mov':
    Duration: 00:01:09.0, start: 0.000000, bitrate: 28847 kb/s
    Stream #0.0(eng): Audio: aac, 44100 Hz, stereo
    Stream #0.1(eng), 59.75 fps(r): Video: dvvideo, yuv411p, 720x480
  Stream mapping:
    Stream #0.1 -> #0.0
    Stream #0.0 -> #0.1
  Press [q] to stop encoding"
  
  @empty_file = "ERROR: undefined method `timestamp' for nil:NilClass ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flv/stream.rb:285:in `lasttimestamp' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flv/stream.rb:274:in `duration' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:181:in `add_meta_data_tag' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:137:in `update' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:47:in `send' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:47:in `execute!' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:46:in `each' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:46:in `execute!' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:239:in `process_files' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:225:in `each' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:225:in `process_files' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2/base.rb:44:in `execute!' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2.rb:168:in `execute!' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/lib/flvtool2.rb:228 ERROR: /usr/lib/ruby/1.8/rubygems/custom_require.rb:27:in `gem_original_require' ERROR: /usr/lib/ruby/1.8/rubygems/custom_require.rb:27:in `require' ERROR: /var/lib/gems/1.8/gems/flvtool2-1.0.6/bin/flvtool2:2 ERROR: /var/lib/gems/1.8/bin/flvtool2:18:in `load' ERROR: /var/lib/gems/1.8/bin/flvtool2:18"
  
  @metadata_result = "---
  /Users/jon/code/spinoza/rvideo/temp.flv: 
    hasKeyframes: true
    cuePoints: 
    audiodatarate: 64.8512825785226
    hasVideo: true
    stereo: false
    canSeekToEnd: false
    framerate: 30
    audiosamplerate: 44000
    videocodecid: 2
    datasize: 992710
    lasttimestamp: 19.453
    audiosamplesize: 16
    audiosize: 165955
    hasAudio: true
    audiodelay: 0
    videosize: 825165
    metadatadate: Fri Sep 14 13:25:58 GMT-0500 2007
    metadatacreator: inlet media FLVTool2 v1.0.6 - http://www.inlet-media.de/flvtool2
    lastkeyframetimestamp: 19.219
    height: 240
    filesize: 998071
    hasMetadata: true
    keyframes: 
      times: 
        - 0
        - 0.4
        - 0.801
        - 1.201
        - 1.602
        - 2.002
        - 2.402
        - 2.803
        - 3.203
        - 3.604
        - 4.004
        - 4.404
        - 4.805
        - 5.205
        - 5.606
        - 6.006
        - 6.406
        - 6.807
        - 7.207
        - 7.608
        - 8.008
        - 8.408
        - 8.809
        - 9.209
        - 9.61
        - 10.01
        - 10.41
        - 10.811
        - 11.211
        - 11.612
        - 12.012
        - 12.412
        - 12.813
        - 13.213
        - 13.614
        - 14.014
        - 14.414
        - 14.815
        - 15.215
        - 15.616
        - 16.016
        - 16.416
        - 16.817
        - 17.217
        - 17.618
        - 18.018
        - 18.418
        - 18.819
        - 19.219
      filepositions: 
        - 1573
        - 24627
        - 56532
        - 90630
        - 137024
        - 185134
        - 225110
        - 262990
        - 291508
        - 330947
        - 370739
        - 398621
        - 426203
        - 448515
        - 468895
        - 488660
        - 508208
        - 523991
        - 541463
        - 558463
        - 572248
        - 590480
        - 604788
        - 620959
        - 632658
        - 645806
        - 655949
        - 668266
        - 684172
        - 697064
        - 713306
        - 728607
        - 746615
        - 760836
        - 774867
        - 790766
        - 804779
        - 821676
        - 843681
        - 857278
        - 873427
        - 888404
        - 900958
        - 914336
        - 927537
        - 941037
        - 958188
        - 974841
        - 988796
    audiocodecid: 2
    videodatarate: 336.705289672544
    duration: 19.486
    hasCuePoints: false
    width: 320
  ..."
end

spec/units/transcoder_spec.rb
require File.dirname(__FILE__) + '/../spec_helper'

def setup_spec
  @options = {:output_file => "bar", :resolution => "baz"}
  @input_file = "#{TEST_FILE_PATH}/kites.mp4"
  @simple_avi = "ffmpeg -i $input_file$ -ar 44100 -ab 64 -vcodec xvid -acodec mp3 -r 29.97 -s $resolution$ -y $output_file$"
  @transcoder = RVideo::Transcoder.new(@input_file)
  @mock_original_file = mock(:original)
  @mock_original_file.stub!(:raw_response)
  RVideo::Inspector.stub!(:new).and_return(@mock_original_file)
end

module RVideo
  
  describe Transcoder, " execution" do
    before do
      setup_spec
      @transcoder.stub!(:check_integrity).and_return(true)
    end
    
    it "should pass a string as-is, along with options" do
      @transcoder.stub!(:parse_and_execute)
      @simple_avi = "ffmpeg -i foo"
      @transcoder.execute(@simple_avi, @options)
    end
    
    it "should store a Tool object at transcoder.executed_commands" do
      @mock_ffmpeg = mock("ffmpeg")
      Tools::AbstractTool.stub!(:assign).and_return(@mock_ffmpeg)
      @mock_ffmpeg.should_receive(:execute)
      @mock_ffmpeg.stub!(:original=)
      @transcoder.execute(@simple_avi, @options)
      @transcoder.executed_commands.size.should == 1
      @transcoder.executed_commands.first.should == @mock_ffmpeg
    end
    
    it "should set original file" do
      @mock_ffmpeg = mock("ffmpeg")
      Tools::AbstractTool.stub!(:assign).and_return(@mock_ffmpeg)
      @mock_ffmpeg.stub!(:execute)
      @mock_ffmpeg.should_receive(:original=).with(@transcoder.original)
      @transcoder.execute(@simple_avi, @options)
    end
    
    it "should raise an exception when trying to call a tool that doesn't exist" do
      lambda {
        @transcoder.send(:parse_and_execute, "foo -i bar", {})
      }.should raise_error(TranscoderError::UnknownTool, /recipe tried to use the 'foo' tool/)
    end
    
    it "should raise an exception when the first argument is not a string" do
      [String, 1, 1.0, true, nil, :foo].each do |obj|
        lambda {
          @transcoder.execute(obj, @options)
        }.should raise_error(TranscoderError::ParameterError, /expected.*recipe.*string/i)      
      end
    end
  end
  
  describe Transcoder, " file integrity checks" do

    before do
      setup_spec
      @transcoder.stub!(:parse_and_execute)
      @mock_processed_file = mock("processed")
      @mock_original_file.stub!(:duration).and_return 10
      @mock_original_file.stub!(:invalid?).and_return false
      @mock_processed_file.stub!(:duration).and_return 10
      @mock_processed_file.stub!(:invalid?).and_return false
      #Inspector.should_receive(:new).once.with(:file => "foo").and_return(@mock_original_file)
      Inspector.should_receive(:new).once.with(:file => "bar").and_return(@mock_processed_file)
    end

    it "should call the inspector twice on a successful job, and should set @original and @processed" do
      @transcoder.original.should_not be_nil
      @transcoder.processed.should be_nil

      @transcoder.execute(@simple_avi, @options)
      @transcoder.original.should == @mock_original_file
      @transcoder.processed.should == @mock_processed_file
    end
    
    it "should check integrity" do
      @transcoder.should_receive(:check_integrity).once.and_return true
      @transcoder.execute(@simple_avi, @options).should be_true
      @transcoder.errors.should be_empty
    end
    
    it "should fail if output duration is more than 10% different than the original" do
      @mock_original_file.should_receive(:duration).twice.and_return(10)
      @mock_processed_file.should_receive(:duration).twice.and_return(13)
      @transcoder.execute(@simple_avi, @options).should be_false
      @transcoder.errors.should == ["Original file has a duration of 10, but processed file has a duration of 13"]
    end
    
    it "should fail if the processed file is invalid" do
      @mock_processed_file.should_receive(:invalid?).and_return(true)
      @transcoder.execute(@simple_avi, @options).should be_false
      @transcoder.errors.should_not be_empty
    end
    
  end
  
  describe Transcoder, "#parse_and_execute" do
    before do
      setup_spec
      @mock_tool = mock("tool")
      @mock_tool.stub!(:execute)
      @mock_tool.stub!(:original=)
      @options_plus_input = @options.merge(:input_file => @input_file)
    end
    
    it "should assign a command via AbstractTool.assign, and pass the right options" do
      Tools::AbstractTool.should_receive(:assign).with(@simple_avi, @options_plus_input).and_return(@mock_tool)
      @transcoder.send(:parse_and_execute, @simple_avi, @options)
    end

    it "should call Tools::AbstractTool once with a one-line recipe" do
      Tools::AbstractTool.should_receive(:assign).once.and_return(@mock_tool)
      @transcoder.send(:parse_and_execute, @simple_avi, @options)
    end
    
    it "should call twice with a two-line recipe" do
      two_line = "ffmpeg -i foo \n ffmpeg -i bar"
      Tools::AbstractTool.should_receive(:assign).twice.and_return(@mock_tool)
      @transcoder.send(:parse_and_execute, two_line, @options)
    end
    
    it "should call five times with a five-line recipe" do
      five_line = "ffmpeg -i foo \n ffmpeg -i bar \n flvtool -i foo \n mp4box \n qt tools 8"
      Tools::AbstractTool.should_receive(:assign).exactly(5).and_return(@mock_tool)
      @transcoder.send(:parse_and_execute, five_line, @options)
    end
    
    it "should pass an exception from the abstract tool" do
      Tools::AbstractTool.should_receive(:assign).and_raise(TranscoderError::UnexpectedResult)
      
      lambda {
        @transcoder.send(:parse_and_execute, @simple_avi, @options)
      }.should raise_error(TranscoderError::UnexpectedResult)
    end
  end

end


lib/rvideo/tools/ffmpeg.rb
module RVideo
  module Tools
    class Ffmpeg
      include AbstractTool::InstanceMethods
      
      attr_reader :frame, :q, :size, :time, :output_bitrate, :video_size, :audio_size, :header_size, :overhead, :psnr, :output_fps
      
      # Not sure if this is needed anymore...
      def tool_command
        'ffmpeg'
      end
      
      #
      # Return -r and the frame rate of the original file. E.g.:
      #
      #   -r 29.97
      #
      # If the original frame rate can't be determined, return an empty 
      # string.
      def original_fps
        inspect_original if @original.nil?
        if @original.fps
          "-r #{@original.fps}" 
        else
          ""
        end
      end
      
      private
      
      def parse_progress(line, duration)
        if line =~ /Duration: (\d{2}):(\d{2}):(\d{2}).(\d{1})/
          duration = (($1.to_i * 60 + $2.to_i) * 60 + $3.to_i) * 10 + $4.to_i
        end
        if line =~ /time=(\d+).(\d+)/
          if not duration.nil? and duration != 0
            p = ($1.to_i * 10 + $2.to_i) * 100 / duration
          else
            p = 0
          end
          p = 100 if p > 100
          return [p, duration]
        end
        nil
      end
      
      def parse_result(result)
        
        if m = /Unable for find a suitable output format for.*$/.match(result)
          raise TranscoderError::InvalidCommand, m[0]
        end
        
        if m = /Unknown codec \'(.*)\'/.match(result)
          raise TranscoderError::InvalidFile, "Codec #{m[1]} not supported by this build of ffmpeg"
       end
        
        if m = /could not find codec parameters/.match(result)
          raise TranscoderError::InvalidFile, "Codec not supported by this build of ffmpeg"
        end
        
        if m = /I\/O error occured\n(.*)$/.match(result)
          raise TranscoderError::InvalidFile, "I/O error: #{m[1].strip}"
        end
          
        if m = /\n(.*)Unknown Format$/.match(result)
          raise TranscoderError::InvalidFile, "unknown format (#{m[1]})"
        end
        
        if m = /\nERROR.*/m.match(result)
          raise TranscoderError::InvalidFile, m[0]
        end
        
        if result =~ /usage: ffmpeg/
          raise TranscoderError::InvalidCommand, "must pass a command to ffmpeg"
        end
        
        if result =~ /Output file does not contain.*stream/
          raise TranscoderError, "Output file does not contain any video or audio streams."
        end
        
        if m = /Unsupported codec.*id=(.*)\).*for input stream\s*(.*)\s*/.match(result) 
          inspect_original if @original.nil?
          case m[2]
          when @original.audio_stream_id
            codec_type = "audio"
            codec = @original.audio_codec
          when @original.video_stream_id
            codec_type = "video"
            codec = @original.video_codec
          else
            codec_type = "video or audio"
            codec = "unknown"
          end
          
          raise TranscoderError::InvalidFile, "Unsupported #{codec_type} codec: #{codec} (id=#{m[1]}, stream=#{m[2]})"
          #raise TranscoderError, "Codec #{m[1]} not supported (in stream #{m[2]})"
        end
        
        # Could not open './spec/../config/../tmp/processed/1/kites-1.avi'
        if result =~ /Could not open .#{@output_file}.\Z/
          raise TranscoderError, "Could not write output file to #{@output_file}"
        end
          
        full_details = /Press .* to stop encoding\n(.*)/m.match(result)
        raise TranscoderError, "Unexpected result details (#{result})" if full_details.nil?
        details = full_details[1].strip.gsub(/\s*\n\s*/," - ")
        
        if details =~ /Could not write header/
          raise TranscoderError, details
        end
        
        #frame=  584 q=6.0 Lsize=     708kB time=19.5 bitrate= 297.8kbits/s    
        #video:49kB audio:153kB global headers:0kB muxing overhead 250.444444%
        
        #frame= 4126 q=31.0 Lsize=    5917kB time=69.1 bitrate= 702.0kbits/s    
        #video:2417kB audio:540kB global headers:0kB muxing overhead 100.140277%
        
        #frame=  273 fps= 31 q=10.0 Lsize=     398kB time=5.9 bitrate= 551.8kbits/s
        #video:284kB audio:92kB global headers:0kB muxing overhead 5.723981%
        
        #mdb:94, lastbuf:0 skipping granule 0
        #size=    1080kB time=69.1 bitrate= 128.0kbits /s    
        #video:0kB audio:1080kB global headers:0kB muxing overhead 0.002893%
        
        #size=      80kB time=5.1 bitrate= 128.0kbits/s    ^Msize=     162kB time=10.3 bitrate= 128.0kbits/s    ^Msize=     241kB time=15.4 bitrate= 128.0kbits/s    ^Msize=     329kB time=21.1 bitrate= 128.0kbits/s    ^Msize=     413kB time=26.4 bitrate= 128.0kbits/s    ^Msize=     506kB time=32.4 bitrate= 128.0kbits/s    ^Msize=     591kB time=37.8 bitrate= 128.0kbits/s    ^Msize=     674kB time=43.2 bitrate= 128.0kbits/s    ^Msize=     771kB time=49.4 bitrate= 128.0kbits/s    ^Msize=     851kB time=54.5 bitrate= 128.0kbits/s    ^Msize=     932kB time=59.6 bitrate= 128.0kbits/s    ^Msize=    1015kB time=64.9 bitrate= 128.0kbits/s    ^Msize=    1094kB time=70.0 bitrate= 128.0kbits/s    ^Msize=    1175kB time=75.2 bitrate= 128.0kbits/s    ^Msize=    1244kB time=79.6 bitrate= 128.0kbits/s    ^Msize=    1335kB time=85.4 bitrate= 128.0kbits/s    ^Msize=    1417kB time=90.7 bitrate= 128.0kbits/s    ^Msize=    1508kB time=96.5 bitrate= 128.0kbits/s    ^Msize=    1589kB time=101.7 bitrate= 128.0kbits/s    ^Msize=    1671kB time=106.9 bitrate= 128.0kbits/s    ^Msize=    1711kB time=109.5 bitrate= 128.0kbits/s - video:0kB audio:1711kB global headers:0kB muxing overhead 0.001826%
        
        #mdb:14, lastbuf:0 skipping granule 0 - overread, skip -5 enddists: -2 -2 - overread, skip -5 enddists: -2 -2 - size=      90kB time=5.7 bitrate= 128.0kbits/s    \nsize=     189kB time=12.1 bitrate= 128.0kbits/s
        
        #size=      59kB time=20.2 bitrate=  24.0kbits/s    \nsize=     139kB time=47.4 bitrate=  24.0kbits/s    \nsize=     224kB time=76.5 bitrate=  24.0kbits/s    \nsize=     304kB time=103.7 bitrate=  24.0kbits/s    \nsi
        
        #mdb:14, lastbuf:0 skipping granule 0 - overread, skip -5 enddists: -2 -2 - overread, skip -5 enddists: -2 -2 - size=      81kB time=10.3 bitrate=  64.0kbits/s    \nsize=     153kB time=19.6 bitrate=  64.0kbits/s 
        
        #size=      65kB time=4.1 bitrate= 128.1kbits/s    \nsize=     119kB time=7.6 bitrate= 128.0kbits/s    \nsize=     188kB time=12.0 bitrate= 128.0kbits/s    \nsize=     268kB time=17.1 bitrate= 128.0kbits/s    \nsize=
        
        #Error while decoding stream #0.1     [mpeg4aac @ 0xb7d089f0]faac: frame decoding failed: Gain control not yet implementedError while decoding stream #0.1frame= 2143 fps= 83 q=4.0 size=    4476kB time=71.3 bitrate= 514.5kbits/s    ^M[mpeg4aac @ 0xb7d089f0]faac: frame decoding failed: Gain control not yet implementedError while decoding stream #0.1
        
        # NOTE: had to remove "\s" from "\s.*L.*size=" from this regexp below.
        # Not sure why. Unit tests were succeeding, but hand tests weren't.
        if details =~ /video:/ 
          #success = /^frame=\s*(\S*)\s*q=(\S*).*L.*size=\s*(\S*)\s*time=\s*(\S*)\s*bitrate=\s*(\S*)\s*/m.match(details)
          @frame = sanitary_match(/frame=\s*(\S*)/, details)
          @output_fps = sanitary_match(/fps=\s*(\S*)/, details)
          @q = sanitary_match(/\s+q=\s*(\S*)/, details)
          @size = sanitary_match(/size=\s*(\S*)/, details)
          @time = sanitary_match(/time=\s*(\S*)/, details)
          @output_bitrate = sanitary_match(/bitrate=\s*(\S*)/, details)
          
          @video_size = /video:\s*(\S*)/.match(details)[1]
          @audio_size = /audio:\s*(\S*)/.match(details)[1]
          @header_size = /headers:\s*(\S*)/.match(details)[1]
          @overhead = /overhead[:]*\s*(\S*)/.match(details)[1]
          psnr_match = /PSNR=(.*)\s*size=/.match(details)
          @psnr = psnr_match[1].strip if psnr_match
          return true
        end
        
        #[mp3 @ 0x54340c]flv doesnt support that sample rate, choose from (44100, 22050, 11025)
        #Could not write header for output file #0 (incorrect codec parameters ?)
        
        raise TranscoderError::UnexpectedResult, details
      end

      def sanitary_match(regexp, string)
        match = regexp.match(string)
        return match[1] if match
      end
      
    end
  end
end
